example/convolution1D/src/convolution1D.cpp

```diff
@@ -0,0 +1,200 @@
+/* Copyright 2023 Benjamin Worpitz, Erik Zenker, Bernhard Manfred Gruber, Jan Stephan
+ * SPDX-License-Identifier: ISC
+ */
+
+#include <alpaka/alpaka.hpp>
+#include <alpaka/example/ExampleDefaultAcc.hpp>
+
+#include <cmath>
+#include <iomanip>
+#include <iostream>
+#include <limits>
+#include <type_traits>
+
+//! Convolution Example
+//!
+//! 1D convolution example: Creates two 1D arrays, applies convolution filter.
+//! Array sizes are hardcoded.
+//!
+
+/**
+ * @brief The ConvolutionKernel function-object
+ * Calculates 1D convolution using input and filter arrays.
+ */
+struct ConvolutionKernel
+{
+    /** @brief Main convolution code
+     *  @param Accelerator
+     *  @param Input array, first input of convolution integral
+     *  @param Filter array, second input of convolution integral
+     *  @param Empty output array to be filled
+     *  @param Input array size
+     *  @param Filter size
+     */
+    template<typename TAcc, typename TElem>
+    ALPAKA_FN_ACC auto operator()(
+        TAcc const& acc,
+        TElem const* const input,
+        TElem const* const filter,
+        TElem* const output,
+        const std::size_t inputSize,
+        const std::size_t filterSize) const -> void
+    {
+        using Idx = alpaka::Idx<TAcc>;
+        using Vec = alpaka::Vec<alpaka::Dim<TAcc>, Idx>;
+        using VecInt = alpaka::Vec<alpaka::DimInt<1u>, Idx>;
+
+
+        Vec const globalThreadIdx = alpaka::getIdx<alpaka::Grid, alpaka::Threads>(acc);
+        Vec const globalThreadExtent = alpaka::getWorkDiv<alpaka::Grid, alpaka::Threads>(acc);
+
+        // Since the kernel is launched 1-D calculating linearizedGlobalThreadIdx line is unnecessary.
+        // globalThreadIdx[0] can be used to map all the threads.
+        if(globalThreadIdx[0] < static_cast<unsigned>(inputSize))
+        {
+            uint32_t const halfFilterSize = filterSize / 2;
+            TElem result = 0.0f;
+            // Calculate sum of multiplications of corresponding elements
+            for(size_t i = 0; i < filterSize; ++i)
+            {
+                uint32_t inputIndex = globalThreadIdx[0] - halfFilterSize + i;
+                if(inputIndex >= 0 && inputIndex < inputSize)
+                {
+                    result += input[inputIndex] * filter[i];
+                }
+            }
+            output[globalThreadIdx[0]] = result;
```

Comment:
I would avoid rechecking the condition inside the loop and ajust the loop range. Also, your code contains an underflow in (`globalThreadIdx[0] - halfFilterSize` may be below 0, which would then be a huge number, but the `+ i` would then overflow again), which works by accident because `uint32_t` has defined wrap-around semantics, and `inputSize` is not close to `UINT_MAX`.
```suggestion
            uint32_t const halfFilterSize = filterSize / 2;
            TElem result = 0.0f;
            // Calculate sum of multiplications of corresponding elements
            auto const start = static_cast<uint32_t>(std::max(static_cast<int>(globalThreadIdx[0]) - halfFilterSize, 0));
            auto const stop = std::min(globalThreadIdx[0] + halfFilterSize, inputSize);
            for(size_t i = start; i <= stop; ++i)
                    result += input[i] * filter[i-globalThreadIdx[0]];
```
This may not be the nicest version and you could consider still letting the loop run from 0 to `filterSize` and offset the index into `input` instead.

Reply:
yes, i had that in mind first. Then in the last commit today, I converted int to uint32_t by mistake after seeing "use types with explicit sizes". I will use int32_t.

Reply:
I think it is ok now. Thanks good point.

Reply:
About checking condition inside the loop; is it because of kernel code or just for optimization in general?

Reply:
> I will simply use int32_t.

You can also just use `int`.

Reply:
> About checking condition inside the loop: is it because this is kernel code ( warp coalescence etc) or just for optimization in general?

General optimization.

Reply:
int is not allowed in [coding guidelines ](https://alpaka.readthedocs.io/en/latest/dev/style.html)of alpaka.

Reply:
Hmm, I think the coding guidelines need a rewrite then. I think there is nothing wrong with using an `int` as a loop counter or distance measure.

Reply:
#2231

Reply:
Inside the new loop you suggested, (i - globalThreadIndx(0) ) becomes  -1.  I needed another check in the loop. Another solution is padding, I did it in the 2d example.

Reply:
> Inside the new loop you suggested, (i - globalThreadIndx(0) ) becomes -1.

Right, because I forgot a `+halfFilterSize` and a `-1` :D Consider the code above (and below) pseudocode. Try:
```c++
            auto const start = static_cast<uint32_t>(std::max(static_cast<int>(globalThreadIdx[0]) - halfFilterSize, 0));
            auto const stop = std::min(globalThreadIdx[0] + halfFilterSize, inputSize-1);
            for(size_t i = start; i <= stop; ++i)
                    result += input[i] * filter[i+halfFilterSize-globalThreadIdx[0]];
```
> I needed another check in the loop.

It should not be necessary if the loop bounds are correct.

> Another solution is padding, I did it in the 2d example therefore this can be left like this.

Right, which may be a better solution, since it makes the kernel simpler.